File: leetcode/algorithms/448_find_all_numbers_disappeared_in_an_array/main.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class FindAllNumbersDisappearedInAnArray {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        List<Integer> result = new ArrayList<Integer>();
        int length = nums.length;
        nums = Arrays.stream(nums).distinct().sorted().toArray();
        int count = 0;
        for (int i = 1; i <= length; i++) {
            if (count >= nums.length - 1) {
                result.add(i);
                continue;
            }
            if (nums[count] != i) {
                result.add(i);
            } else {
                count++;
            }
        }
        return result;
    }


    // Best Solution
    public List<Integer> bestSolution(int[] nums) {
        List<Integer> list = new ArrayList<>();
        int idx = -1;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] < 0) {
                idx = nums[i] * -1 - 1;
            } else {
                idx = nums[i] - 1;
            }

            if (nums[idx] > 0) {
                nums[idx] = -nums[idx];
            }

        }

        for (int i = 0; i < nums.length; i++) {
            if (nums[i] > 0) {
                list.add(i + 1);
            }
        }

        return list;
    }
}

File: leetcode/algorithms/448_find_all_numbers_disappeared_in_an_array/main.dart
class FindAllNumbersDisappearedInAnArray {
  List<int> findDisappearedNumbers(List<int> nums) {
    List<int> numArray = List<int>.filled(nums.length, 0);

    for (int num in nums) {
      numArray[num - 1] = 1;
    }

    List<int> result = List<int>.empty(growable: true);

    for (int i = 1; i <= nums.length; i++) {
      if (numArray[i - 1] == 0) {
        result.add(i);
      }
    }

    return result;
  }

  // Best Solution
  List<int> bestSolution(List<int> nums) {
    Set<int> distinct = {...nums};
    int len = nums.length;
    List<int> res = [];
    for (int i = 1; i <= len; i++) {
      if (!(distinct.contains(i))) {
        res.add(i);
      }
    }
    return res;
  }
}
